### DISCUSSION TITLE
How to map extern variables to specific ZP locations

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1361] [Q&A] [answered]
I am trying to have some variables in ZP.
In my .c files I have:
```
extern uint8_t snake_head_x;
#if defined(__CC65__)
#pragma zpsym("snake_head_x")
#endif
extern uint8_t snake_head_y;
#if defined(__CC65__)
#pragma zpsym("snake_head_y")
#endif
```

I compile and link with a .s file which contains:
```
.exportzp _snake_x;
_snake_x = $2B
.exportzp _snake_y;
_snake_y = $2C 
```
(I have also tried with `.export` without any change)

I get:
```
ld65: Warning: Address size mismatch for '_snake_x': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(11) as 'zeropage', import in move_snake.o, games/snake/move_snake.s(17) as 'absolute'
ld65: Warning: Address size mismatch for '_snake_x': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(11) as 'zeropage', import in snake.o, games/snake/snake.s(20) as 'absolute'
ld65: Warning: Address size mismatch for '_snake_x': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(11) as 'zeropage', import in main.o, games/snake/main.s(80) as 'absolute'
ld65: Warning: Address size mismatch for '_snake_y': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(14) as 'zeropage', import in move_snake.o, games/snake/move_snake.s(18) as 'absolute'
ld65: Warning: Address size mismatch for '_snake_y': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(14) as 'zeropage', import in snake.o, games/snake/snake.s(21) as 'absolute'
ld65: Warning: Address size mismatch for '_snake_y': Exported from vic20_memory.o, games/snake/memory/vic20_
memory.s(14) as 'zeropage', import in main.o, games/snake/main.s(81) as 'absolute'
ld65: Error: Cannot generate most of the files due to memory area overflow
make: *** [Makefile.snake:419: vic20_exp_8k] Error 1
```

Whay is wrong in what I am doing?

**Comment by Fabrizio-Caruso** [comment]
I figure it out... There is a stupid typo/bug.
Is `exportzp `really necessary? A plain `export` seems to work equally well.

**Comment by oliverschmidt** [comment]
* It seems to me there's something wrong with your example - the .s file is about `snake_x` while your .c file is about `snake_head_x`.

* If you paste cc65 tool message referring to files like `vic20_memory.s`, `move_snake.s` or `main.s` you need to supply the information which code snippets given are part of which file.

**Comment by Fabrizio-Caruso** [comment]
You are right. It is a stupid error on my side.
There are still things I don't understand, though.
1. `export` seems to work as `exportzp`. I must be missing some important detail here.
2. Do I have to use zpsym on any .c file that uses the very extern variable in ZP? Can I just the pragma once?

**Comment by oliverschmidt** [comment]
Regarding 2.) - I'd phrase it this way: The `#pragma zpsym()` belongs everywhere, where the `extern` belongs - which is usually in a .h file.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt Thanks for the advise. I am doing the wrong thing.
I have all my .c files polluted with `extern` declarations. 
I have to clean up my mess.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt  I am not sure I know how to use your good advise in my use-case.
My game-code is supposed to be hardware-agnostic and dev-kit agnostic. 
It is compiled with 12 different cross-compilers. The game code is not supposed to change ever but can be linked to other target-specific files and also uses some macros that depend on options in order to do some magic.
The goal is to hide as much as possible any dev-kit dependency and keep the code hardware-agnostic.

In particular for some CC65 targets (and Z88DK targets) I want to map some variables specific memory locations.
So far, I have done this by having all .c use extern declarations.
I have put all variables in one single file where I define _EXTERN as either empty string or "extern" when I want to link it to an Assembly file to import some variables instead of defining them.
I guess I would have to move all variables definitions into one globals.c file.

**Comment by Fabrizio-Caruso** [comment]
I am not I have been able to clearly explain what I am doing with my main project (Cross-Lib)...

The user of Cross-Lib should write a C program that only depends on Cross-Lib APIs and plain C.
Then the build process of my framework will do "the magic", which includes compiling and linking it with Cross-Lib and even target-specific files.

How can I use CC65-specific pragmas in CC65-agnostic game code?
The current dirty solution for optional non-ZP memory exports has been to export all variables in the whole game code except for a variables.h file, which either defines the variables or export them when also an Assembly  (CC65 or Z88DK) file is linked.
This allows me to map variables to memory locations while still keeping the game code devkit-agnostic with the sole exception of the Assembly file to be linked.
My current solution breaks if I want to use ZP with CC65.... This is a problem with my approach and not with CC65. I am just asking for some advice..

**Comment by Fabrizio-Caruso** [comment]
Moreover I don't think I can create a macro that embeds a pragma and be sure all 12 cross-compilers will behave nicely with it....

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I have followed you advice and I have centralized all extern declarations in a single header file.

In order to be able to switch on a dev-kit and target base whether a target has to use an Assembly file to map variables onto memory, I have centralzed all global variable definitions into one single file. This seems ugly and it is slightly less memory efficient for some reason I do not fully understand.  It allows me to switch between a normal definition and an extern declaration with possible additional zp pragma directive:

https://github.com/Fabrizio-Caruso/CROSS-LIB/tree/master/src/games/snake

(Fourth universal Cross-Lib game)

**Comment by oliverschmidt** [comment, marked as answer]
> This seems ugly [...]

Every variable *declaration* belongs into a .h files while every variable *definition* belongs into exactly one .c file (or .s file). That doesn't mean at all that *all* variables must be declared in a single .h file nor does it mean that *all* variables must be defined in a single .c file.

The usual pattern ist that one "thing" in your program "owns" some variables and that the other "things" in your program use those variables. So you create e.g. `snake.c` defining the snake variables (and defining the snake functions) and `snake.h` declaring those snake variables (and declaring those snake functions). You then include `snake.h` both in `snake.c` and all .c file using the snake variables and/or the snake functions.

Note 1: Including `snake.h` in `snake.c` is done to allow the compiler to check if the declaration of the variables (and functions) match the definitions of the variables (and functions).

Note 2: You'll end up with quite some .h files (usually one per .c file). Some people consider it great that a usual .c therefore contains quite some `#include` statements as they show which parts of the program are going to be used in that .c file. Yet other people rather create a "master include" file: A .h file just consisting of `#include` statements of most/all other .h files. Call it a matter of taste.

> [...] it is slightly less memory efficient [...]

No. The linker pulls all the variable definitions from all .c files together into the same segment. It doesn't matter how the definitions are distributed over the files.

**Reply by Fabrizio-Caruso** [reply]
You are right about the ugliness.
I will try to improve it.
The problem is that I do not want to insert anything that is specific to CC65 (or Z88DK or any other dev-kit) in the game code, which in my case should be dev-kit agnostic.
I have to find a way to hide with a macro so that in the file xxx.c I can have something that will declare OR define a variables depending on a compiler option.

**Comment by oliverschmidt** [comment]
> I can have something that will declare OR define a variables depending on a compiler option.

Why can't you just always define in one place (either just in C or via cc65 .s file) and always declare in another place. To me it seems a bad idea (tm) to switch between definition and declaration via macro.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt the reason is what I tried to explain above, i.e., I don't want the game code to have any CC65-specific pragmas as the game code is supposed to be compiled also by Z88DK, LC1802, ACK, CMOC, GCC (several modded and native versions), XTC68, etc...
The very goal of my project is to have dev-kit-agnostic APIs. You write you little game once in C + my Cross-Lib APIs and you compile it and build it through my scripts and makefiles to get 200 versions for any 8-bit architectures with even some simple sounds and graphics. Even the assets are converted on the fly for any dev-kit and target.
A similar project to mine is 8-bit-Unity (which has more features but only covers 5 CC65 targets whereas I cover about 200, over all 8-bit architectures).
This is why some time ago I asked to have an option that would just place globals in ZP via something other than pragmas in the code.

I agree with you on the fact that my current solution is ugly and I want to improve it.
